added support for sending telemetry to eventhubs

martin-weber opened the pull request:
- added EventHubConnectionString command line parameter to IotTelemetrySimulator
- moved IoT Hub and Event Hub specific code out of SimulatedDevice into implementations of ISender to make SimulatedDevices agnostic of the message sink
- moved IotHub DeviceClient out of Factory int ISender implementation of the corresponding message sink

martin-weber committed:
added support for sending telemetry to eventhubs

- added EventHubConnectionString parameter to sender
- moved iot hub and event hub specific code out of  into implementations of ISender to make SimulatedDevices agnostic of the message sink.

fbeltrao commented on src/IotTelemetrySimulator/EventHubSender.cs:
```diff
@@ -0,0 +1,99 @@
+                try
+                {
+                    await eventHubClient.SendAsync(msg);
+                    stats.IncrementMessageSent();
```
We are repeating the `stats.IncrementMessageSent()` and `stats.IncrementSendTelemetryErrors()` in both senders. Wonder if we can share this code in a common class (i.e. composition pattern)

martin-weber replied:
I added an abstract base class SenderBase to reduce code duplication

fbeltrao commented on src/IotTelemetrySimulator/DefaultDeviceSimulatorFactory.cs:
```diff
@@ -1,26 +1,25 @@
+
+
+            if (!string.IsNullOrEmpty(config.EventHubConnectionString))
+                return new EventHubSender(deviceId, config);
```
We should reuse the EventHubClient here

martin-weber replied:
Agree. Done

fbeltrao reviewed (commented):
@martin-weber Thanks for contribution. Sorry for the delay. Finally had the time to look at it.

martin-weber committed:
refactored according to PR feedbacks:
- added abstract SenderBase class to reduce duplicated code
- reuse of the eventHubClient in DefautlDeviceSimulatorFactory

fbeltrao commented:
@martin-weber could you please resolve the conflicts?

martin-weber committed:
rebased on upstream/master

martin-weber commented:
Refactored code to have a common SenderBase which reduced duplicated code to a minimum.

fbeltrao commented on src/IotTelemetrySimulator/SenderBase.cs:
```diff
@@ -0,0 +1,93 @@
+                    stats.IncrementMessageSent();
+                    break;
+                }
+                catch (TCommunicationException)
```
Maybe if we define an virtual/abstract method called `bool IsTransientException(Exception)` we can remove the TCommunicationException. I think it is a bit easier to follow.

martin-weber replied:
Good idea.

fbeltrao commented on src/IotTelemetrySimulator/SenderBase.cs:
```diff
@@ -0,0 +1,93 @@
+
+namespace IotTelemetrySimulator
+{
+    public abstract class SenderBase<TMessage, TCommunicationException> : ISender 
```
Nice one!

martin-weber committed:
changed from generic TCommunicationException to check with when IsTransientException(ex)

martin-weber committed:
added unittests for SenderBase

martin-weber commented:
added unit tests for SenderBase
